File: ABC/ABC301-400/ABC364/F/set/F.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

int main() {
	int N, Q;
	cin >> N >> Q;
	vector<tuple<ll, int, int>> edges(Q);
	for (int i = 0; i < Q; i++) {
		ll C;
		int L, R;
		cin >> L >> R >> C;
		edges[i] = {C, L - 1, R - 1};
	}
	sort(edges.begin(), edges.end());
	set<int> st;
	for (int i = 0; i < N - 1; i++) {
		st.insert(i);
	}
	ll ans = 0;
	for (auto [C, L, R]: edges) {
		int cnt = 0;
		while (true) {
			auto bs = st.lower_bound(L);
			if (bs != st.end() && *bs < R) {
				cnt++;
				st.erase(bs);
			} else {
				break;
			}
		}
		ans += (cnt + 1) * C;
	}
	cout << (st.size() == 0 ? ans : -1) << '\n';
}

File: ABC/ABC301-400/ABC364/F/set/F.py
from sortedcontainers import SortedSet

N, Q = map(int, input().split())
edges = []
for _ in range(Q):
	L, R, C = map(int, input().split())
	edges.append((C, L - 1, R - 1))
edges.sort()
st = SortedSet(list(range(N - 1))) # 頂点 j と j + 1 がつながっていないならば集合に含まれるようにする
ans = 0
for C, L, R in edges:
	# L, …, R - 1 を削除する
	cnt = 0
	while True:
		bs = st.bisect_left(L)
		if bs < len(st) and st[bs] < R:
			cnt += 1
			st.pop(bs)
		else:
			break
	ans += (cnt + 1) * C
print(ans if len(st) == 0 else -1)